src/devices/bus/s100/vectordualmode.cpp

```diff
@@ -0,0 +1,315 @@
+0xFF sync byte. The FDD has 16 hard sectors, but the HDD uses a normal
+soft-sectored drive with a PLL on the controller to emulate 32 hard sectors.
+No abnormal MFM clock bits are used.
+
+https://www.bitsavers.org/pdf/vectorGraphic/hardware/7200-1200-02-1_Dual-Mode_Disk_Controller_Board_Engineering_Documentation_Feb81.pdf
+https://archive.org/details/7200-0001-vector-4-technical-information-sep-82
+
+TODO:
+- HDD support
+- ECC
+
+****************************************************************************/
+
+#include "emu.h"
+
+#include "bus/s100/vectordualmode.h"
+#include "formats/vgi_dsk.h"
+
+static const attotime half_bitcell_size = attotime::from_usec(2);
+
+/* Interleave 8 bits with zeros. abcdefgh -> 0a0b0c0d0e0f0g0h */
+static int deposit8(int data)
+{
+	int d = data;
+	d = ((d & 0xf0) << 4) | (d & 0x0f);
+	d = ((d << 2) | d) & 0x3333;
+	d = ((d << 1) | d) & 0x5555;
+	return d;
+}
+
+static uint16_t mfm_byte(uint8_t data, unsigned int prev_data)
+{
+	const unsigned int ext_data = data | (prev_data << 8);
+	const unsigned int clock = ~(ext_data | (ext_data >> 1));
+	return (deposit8(clock) << 1) | deposit8(ext_data);
+}
+
+static uint8_t unmfm_byte(uint16_t mfm)
+{
+	unsigned int d = mfm;
+	d &= 0x5555;
+	d = ((d >> 1) | d) & 0x3333;
+	d = ((d >> 2) | d) & 0x0f0f;
+	d = ((d >> 4) | d) & 0x00ff;
+	return d;
+}
+
+s100_vector_dualmode_device::s100_vector_dualmode_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, S100_VECTOR_DUALMODE, tag, owner, clock)
+	, device_s100_card_interface(mconfig, *this)
+	, m_floppy(*this, "floppy%u", 0U)
+	, m_ram{0}
+	, m_cmar(0)
+	, m_drive(0)
+	, m_sector(0)
+	, m_read(false)
+	, m_pll()
+	, m_byte_timer(nullptr)
+	, m_sector_timer(nullptr)
+	, m_pending_byte(0)
+	, m_pending_size(0)
+{
+}
+
+TIMER_CALLBACK_MEMBER(s100_vector_dualmode_device::motor_off)
+{
+	for (int i = 0; i < m_floppy.size(); i++) {
+		floppy_image_device* flop = m_floppy[m_drive]->get_device();
+		if (flop)
+			flop->mon_w(1);
+	}
+}
+
+bool s100_vector_dualmode_device::hdd_selected()
+{
+	return m_drive == 0 && false;
+}
+
+uint8_t s100_vector_dualmode_device::s100_sinp_r(offs_t offset)
+{
+	// 7200-1200-02-1 page 16 (1-10)
+	uint8_t data;
+	if (offset == 0xc0) { // status (0) port
+		bool write_protect; // FDD
+		bool ready; // HDD
+		bool track0;
+		bool write_fault = false; // HDD
+		bool seek_complete; // HDD
+		bool loss_of_sync; // HDD
+		if (hdd_selected()) {
+			write_protect = false;
+			ready = true;
+			track0 = false;
+			seek_complete = true;
+			loss_of_sync = true;
+		} else {
+			floppy_image_device* flop = m_floppy[m_drive]->get_device();
+			write_protect = flop && flop->wpt_r();
+			ready = false;
+			track0 = flop && !flop->trk00_r();
+			seek_complete = false;
+			loss_of_sync = false;
+		}
+
+		data = write_protect | (ready << 1) | (track0 << 2)
+		    | (write_fault << 3) | (seek_complete << 4) | (loss_of_sync << 5)
+			| 0xc0;
```

Comment:
Please don’t directly use mathematical operators on `bool` values.  Either turn the variables into `uint8_t` and assign them 0 or 1, or build up the result like `... (ready ? 0x02 : 0x00) | (track0 ? 0x04 : 0x00) ...` etc.  Some compilers and analysis tools produce warnings for using Boolean in integer context.